server/src/main/java/org/opensearch/index/translog/transfer/FileTransferTracker.java

```diff
@@ -85,46 +99,118 @@ void add(String file, boolean success) {
     @Override
-    public void onFailure(TransferFileSnapshot fileSnapshot, Exception e) {
+    public void onFailure(TranslogCheckpointSnapshot fileSnapshot, Exception e) {
         long durationInMillis = (System.nanoTime() - fileTransferStartTime) / 1_000_000L;
         remoteTranslogTransferTracker.addUploadTimeInMillis(durationInMillis);
-        remoteTranslogTransferTracker.addUploadBytesFailed(bytesForTlogCkpFileToUpload.get(fileSnapshot.getName()));
-        add(fileSnapshot.getName(), TransferState.FAILED);
+        addGeneration(fileSnapshot.getGeneration(), TransferState.FAILED);
+
+        if (ckpAsTranslogMetadata) {
+            updateTransferStats(fileSnapshot, false);
+        } else {
+            assert e instanceof TranslogTransferException;
+            TranslogTransferException exception = (TranslogTransferException) e;
+            Set<TransferFileSnapshot> failedFiles = exception.getFailedFiles();
+            Set<TransferFileSnapshot> successFiles = exception.getSuccessFiles();
+            assert failedFiles.isEmpty() == false;
+            failedFiles.forEach(failedFile -> {
+                add(failedFile.getName(), false);
+                long failedBytes = 0;
+                try {
+                    failedBytes = failedFile.getContentLength();
+                } catch (IOException ignore) {}
+                updateBytesInRemoteTranslogTransferTracker(failedBytes, false);
+            });
+            successFiles.forEach(successFile -> {
+                add(successFile.getName(), true);
+                long succededBytes = 0;
+                try {
+                    succededBytes = successFile.getContentLength();
+                } catch (IOException ignore) {}
+                updateBytesInRemoteTranslogTransferTracker(succededBytes, true);
+            });
+        }
     }
 
-    public void delete(List<String> names) {
+    private void updateTransferStats(TranslogCheckpointSnapshot fileSnapshot, boolean isSuccess) {
+        Long translogFileBytes = bytesForTlogCkpFileToUpload.get(fileSnapshot.getTranslogFileName());
+        if (translogFileBytes != null) {
+            updateBytesInRemoteTranslogTransferTracker(translogFileBytes, isSuccess);
+        }
+
+        Long checkpointFileBytes = bytesForTlogCkpFileToUpload.get(fileSnapshot.getCheckpointFileName());
+        if (checkpointFileBytes != null) {
+            updateBytesInRemoteTranslogTransferTracker(checkpointFileBytes, isSuccess);
+        }
+    }
+
+    private void updateBytesInRemoteTranslogTransferTracker(long bytes, boolean isSuccess) {
+        if (isSuccess) {
+            remoteTranslogTransferTracker.addUploadBytesSucceeded(bytes);
+        } else {
+            remoteTranslogTransferTracker.addUploadBytesFailed(bytes);
+        }
+    }
+
+    void delete(List<String> names) {
         for (String name : names) {
             fileTransferTracker.remove(name);
         }
     }
 
+    void deleteGenerations(Set<Long> generations) {
```

Comment:
Where are we using this method?

Reply:
We are using this method to empty the `generationTransferTracker` during the async translog files deletions in `deleteTranslogFilesAsync` function.